Job result Visualizations

annehaley opened the pull request:
This PR introduces client-side visualization of job results. On the backend, we detect nuclei and glands in an image as two separate jobs. The output from either job can be shown in the client by displaying dots over the detected structure centroids. Instead of calculating distance between these centroids as a job on the backend, this type of calculation is better fit for the client, once the client has access to the centroids' information. 

The client does not yet display aggregate information such as the average distance. Right now, the distances are visualized as lines between a nucleus centroid and a gland centroid, appearing when either centroid point is moused over.

Since the backend can detect the presence of either nuclei or glands, the database model of `DetectedNucleus` was generalized as `DetectedStructure` and modified to contain information the client would need.

naglepuff commented on atlascope/core/job_types/gland_detection.py:
```diff
@@ -0,0 +1,107 @@
+
+    try:
+        input_image = skimage.io.imread(io.BytesIO(open(gland_map_path, 'rb').read()))
+        glands, gland_mask = detect_nuclei(input_image)
```
It seems a bit confusing that the output of a function called `detect_nuclei` would be `glands, gland_mask`. Is this function used by both jobs? I think having a separate function for each structure type would make things more clear.

annehaley replied:
Wrote a new `detect_glands` function in [6cbe806](https://github.com/atlascope/atlascope/pull/193/commits/6cbe806fd4df480c24b84b855bcfcf8302a5ef30)

naglepuff commented on atlascope/core/job_types/gland_detection.py:
```diff
@@ -0,0 +1,107 @@
+            centroid = Point(
+                x=gland['Identifier.CentroidX'] * 3,
+                y=gland['Identifier.CentroidY'] * 2.5,
+            )
```
I'm a bit confused by what's happening here. What do you mean by "small sample image?" How were the numbers 3 and 2.5 chosen?

annehaley replied:
We sent the test image `management/populate/inputs/gland_map.png` into this detection. Our test image was much smaller than the associated dataset image, `HTA9_1_BA_F_ROI-2.ome.tiff` (About 3 times smaller horizontally and 2.5 times smaller vertically)  This simple translation of each centroid was merely to compensate for this size difference and move the detected glands to make them more spacially distributed within  `HTA9_1_BA_F_ROI-2.ome.tiff` (rather than isolated in the top right corner, where there are no nuclei).

Instead of further messing with this, I simply wrote a new `detect_glands` method, inspired by your previous comment. Instead of trying to reuse the nucleus detection on a test image, we can just generate where the glands go by generating n dots in a grid within the size of the input image. 

So this is fixed by [6cbe806](https://github.com/atlascope/atlascope/pull/193/commits/6cbe806fd4df480c24b84b855bcfcf8302a5ef30)

naglepuff commented on web/src/components/PinList.vue:
```diff
@@ -55,7 +55,7 @@ import {
 } from '@vue/composition-api';
 import store from '../store';
-import { Pin } from '../generatedTypes/AtlascopeTypes';
+import { Dataset, Pin } from '../generatedTypes/AtlascopeTypes';
```
We don't need this import in this component yet. I get an eslint error with it in.

annehaley replied:
I left this change out when I rebased and reorganized my commits

naglepuff commented on docker-compose.yml:
```diff
@@ -1,8 +1,8 @@
-    logging:
-      driver: none
+    # logging:
+    #   driver: none
```
Should these lines be uncommented?

annehaley replied:
I left this change out when I rebased and reorganized my commits

naglepuff commented on web/src/store/index.ts:
```diff
@@ -198,7 +212,9 @@ const {
-          datasets.forEach((dataset) => {
+          datasets.filter(
+            (dataset) => dataset.dataset_type === 'tile_source',
+          ).forEach((dataset) => {
```
I noticed a need for something like this in one of my branches as well. There's a getter above that can be used here instead of `filter`.

```
const { commit, state, dispatch, getters } = rootActionContext(context);
...
getters.tilesourceDatasets.forEach(...)
```

annehaley replied:
Changed in [5d27ba3](https://github.com/atlascope/atlascope/pull/193/commits/5d27ba37fef2182527ac7f623c2f315f270759a6)

naglepuff commented on web/src/store/index.ts:
```diff
@@ -272,7 +296,26 @@ const {
     updateSelectedPins(context, pins: Pin[]) {
       const { commit } = rootActionContext(context);
-      commit.setSelectedPins(pins);
+      const circlePinDatasetTypes = [
+        'tile_overlay',
+        'non_tiled_image',
+      ];
+      const isVisualization: Array<boolean | Dataset> = pins.map((pin: Pin) => {
+        const childDataset: Dataset | undefined = store.state.currentDatasets.find(
+          (dataset: Dataset) => dataset.id === pin.child,
+        );
+        if (childDataset
+          && childDataset?.dataset_type
+          && !circlePinDatasetTypes.includes(childDataset.dataset_type)) {
+          return childDataset;
+        }
+        return false;
+      });
+      const circlePins = pins.filter((pin: Pin, index: number) => !isVisualization[index]);
+      const visualizations = isVisualization.filter((vis) => vis !== false);
+
+      commit.setSelectedPins(circlePins);
+      commit.setSelectedVisualizations(visualizations);
```
There's certainly a lot happening here now that I'm not sure needs to. I'm wondering if these special visualizations really belong to Pins at all. It doesn't seem like the fit the cork board metaphor of pinning something to the larger dataset, since (at least in the nucleus/glands example) the visualizations apply to the whole of the base dataset.

I think having these just be datasets that get handled specially makes more sense than them being Pins. Let me know if you have any thoughts on this. I think it'd be fine to get this in as a first go but I think there are design considerations that would be worth discussing.

annehaley replied:
I agree, and I took a second pass at this design in #201, though the backend of that design remained unchanged. I do think we should return to this topic.

naglepuff replied:
Looks like #201 was closed in favor of #204. I'm fine with using pins for these two PRs, but I've opened #206 to try and capture some of my feelings with the current design. At this point I'd like to get this feature merged.

naglepuff commented on web/src/utilities/utiltyFunctions.ts:
```diff
@@ -49,3 +51,56 @@ export function opacityForZoomLevel(currentZoom: number, maxZoom: number): numbe
+
+export function centroidStringToCoords(input: string): Array<number> {
+  const centroidString = input.match(/\(([0-9|.|\s])+\)/);
+  if (!centroidString) return [-1, -1];
```
Could you put this line in brackets? It would make it more readable.
```suggestion
  if (!centroidString) { 
    return [-1, -1];
  }
```

annehaley replied:
Changed in [0f3f8b2](https://github.com/atlascope/atlascope/pull/193/commits/0f3f8b2f095924f12167cdf74d4ee27a9ffc0ab4)

naglepuff commented on web/src/utilities/utiltyFunctions.ts:
```diff
@@ -49,3 +51,56 @@ export function opacityForZoomLevel(currentZoom: number, maxZoom: number): numbe
+export interface NucleusGlandDistance{
+  nucleus: number | undefined;
+  gland: number | undefined;
+  line: Array<Array<number>>;
```
Is this always a 2x2 array? Any thoughts on using `Point[]` as the type?

annehaley replied:
Changed in [39c875f](https://github.com/atlascope/atlascope/pull/193/commits/39c875f4db286a66f66b76328637181902f1ce1c)

naglepuff commented on web/src/utilities/analyticsVis.ts:
```diff
@@ -0,0 +1,121 @@
+  x: number;
+  y: number;
+  color: string;
+  struct: DetectedStructure;
```
```suggestion
  structure: DetectedStructure;
```

I think its more clear to just use the full word here.

annehaley replied:
Changed in [515411b](https://github.com/atlascope/atlascope/pull/193/commits/515411be71d623eaac177d2da450d4e18aba311b)

naglepuff reviewed (changes requested):
Really cool to see this in action. As I mentioned in one of my comments, I'm not entirely convinced that Pins are the best way to hook these datasets up to investigations, but I think it's probably ok for now. I had a some other minor comments as well.

annehaley committed:
DetectedNucleus -> DetectedStructure

annehaley committed:
Use gland detection job instead of nucleus to gland distance job

annehaley committed:
Simple jobs accept tiff input instead of png input

annehaley committed:
Fix save_output_dataset function for jobs

annehaley committed:
Change result of populate script for Analytics playground

annehaley committed:
Fix creation of geoJS feature layers to allow passing options

annehaley committed:
Regenerate Atlascope types

annehaley committed:
Separate visualization of pins from visualization of job results

annehaley committed:
Add functions for computing job result visualizations

annehaley committed:
Write new function for detect_glands

Rather than reusing detect_nuclei for gland detecion, we will just
generate n dots in a grid on the image

This way, we don't need to worry about resizing the outputs from a small test image.

annehaley committed:
struct -> structure

naglepuff commented:
Just a heads up, your most recent force push changed the hash for your recent commits, so they don't match the comments you made on the review. I can manage for now but it would be a good idea to update the comments so they are accurate.

annehaley committed:
Use brackets for if statement

annehaley committed:
Use Point[] type instead of number[][]

annehaley commented:
> Just a heads up, your most recent force push changed the hash for your recent commits, so they don't match the comments you made on the review. I can manage for now but it would be a good idea to update the comments so they are accurate.

Nice catch, I updated the links in the resolved issues

annehaley commented:
Apologies for the several force pushes; I'm finding it difficult to merge in changes from main and reconcile the migration files. @mcovalt could you take a look at [3a07b11](https://github.com/atlascope/atlascope/pull/193/commits/3a07b11e36d003f56ca4b08b2c161537ecd5bd2d) and perhaps diagnose why [this test](https://github.com/atlascope/atlascope/runs/7859057521?check_suite_focus=true) failed?

mcovalt commented:
Looks like the index name was autogenerated to be different when on a different model name. I reordered migrations to have my migration last. Let's see if it works. Ideally, we'd not reorder migrations as this will require dropping the DB (or at least rolling back to 008) to reconcile the new migration sequence. You may want to order them such that your migrations occur after the ones on `main`, but up to you.

annehaley commented:
Thanks for fixing that. Last question: Is it possible to supply a null/blank value for the fingerprint field? Now we have glands occupying the same model as nuclei, and I'd like to skip the fingerprint part in the `detect_glands` job.

mcovalt commented:
Looks like the tests are failing due to `fingerprint` not being provided on glands. This exposes a bug in the `CubeField`: a `NOT NULL` constraint should be invalidated immediately, rather than providing what looks like bad data.

Correcting  this behavior in `CubeField` will still result in errors, though. A few options:
- we can allow `null` in the field
- we can just provide a dummy `fingerprint=Cube([-9999]*100)` with glands (probably easiest for now)

In either case, the similar nuclei code assumed `DetectedNuclei` was just nuclei. Expanding its definition to also include glands may break some of the similar nuclei code. I'll keep an eye on this behavior with further development of the similar nuclei features.

annehaley commented:
Ok, we can provide dummy data there. We do so for a lot of the fields on detected glands, since that data is made up. And yes, combining the two models does complicate the logic for your similarity search. If the design of these two features weren't done simultaneously, we may have had time to come up with a better design concept. Something to consider for Phase 2 perhaps.

annehaley committed:
Use dummy fingerprint value for glands

annehaley committed:
Reorder migrations after mergemigrations in main

naglepuff reviewed (approved):
Thanks for getting to all of the feedback and sorting out the migration issues!

waxlamp reviewed (commented):
Looks good. Thanks for wrestling with all of those rebasing and migration challenges!
